src/plugins/console/index.js

```diff
@@ -96,30 +111,64 @@ module.exports = function (kibana) {
         },
 
         pre: [
-          function filterUri(req, reply) {
-            const { uri } = req.query;
+          {
+            assign: 'proxyTarget',
+            method(req, reply) {
+              const rawUri = req.query.uri;
+              req.query.uri = null; // prevent accidental usage
+
+              const { protocol, hostname, port, pathname, search, hash } = parseUri(rawUri);
+              reply(formatUri({
+                protocol,
+                hostname,
+                port,
+                pathname: resolve(pathname), // resolve any path traversal
+                search,
+                hash
+              }));
+            }
+          },
 
-            if (!filters.some(re => re.test(uri))) {
+          function checkProxyTarget(req, reply) {
+            const uri = req.pre.proxyTarget;
+
+            if (!options.proxyTargets) return reply();
+            if (options.proxyTargets.some(t => uri.startsWith(t))) {
+              reply();
+            } else {
               const err = Boom.forbidden();
-              err.output.payload = "Error connecting to '" + uri + "':\n\nUnable to send requests to that url.";
+              err.output.payload = `Error connecting to '${uri}':\n\nUnable to send requests to that url.`;
               err.output.headers['content-type'] = 'text/plain';
               reply(err);
-            } else {
-              reply();
             }
+          },
+
+          function checkProxyFilters(req, reply) {
+            const uri = req.pre.proxyTarget;
+
+            if (filters.some(re => re.test(uri))) {
+              return reply();
+            }
+
+            const err = Boom.forbidden();
+            err.output.payload = `Error connecting to '${uri}':\n\nUnable to send requests to that url.`;
+            err.output.headers['content-type'] = 'text/plain';
+            reply(err);
           }
         ],
 
         handler(req, reply) {
-          const { uri } = req.query;
+          const uri = req.pre.proxyTarget;
 
           reply.proxy({
```

Comment:
This piece should probably use the same header whitelisting logic as in the Kibana proxy. I had implemented that in PR #7067 (which should probably be subsumed by this PR), if you want to steal it from there: https://github.com/ycombinator/kibana/blob/gh-7059/src/plugins/console/index.js#L117-L130